controllers/storageautoscaler/scraper.go

```diff
@@ -0,0 +1,151 @@
+			return
+
+		case <-time.After(r.ScrapeInterval):
+			r.enqueueGenericEventsFromMetrics(ctx)
+		}
+	}
+}
+
+func (r *StorageAutoscalerScraper) enqueueGenericEventsFromMetrics(ctx context.Context) {
+	// query to get the osd usage
+	osdUsageQuery := "(ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))"
+	metrics, err := scrapeMetrics(ctx, r.OperatorNamespace, osdUsageQuery, r.Log)
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return
+	}
+	// refresh the sync map
+	r.SyncMap = &sync.Map{}
+	// update the sync map with the highest value of osd usage per device class
+	updateSyncMap(metrics, r.SyncMap)
+	// get the storage autoscaler objects
+	objectList, err := getStorageAutoScalerObjects(ctx, r.Client)
+	if err != nil {
+		r.Log.Error(err, "failed to get resource objects")
+		return
+	}
+	// filter the objects for scaling
+	filteredObjectList := filterObjectsForScaling(objectList, r.SyncMap, r.Log)
+
+	// send generic events for the filtered objects
+	sendGenericEvent(filteredObjectList, r.EventCh, r.Log)
+}
+
+// update the sync map with the highest value of osd usage per device class
+func updateSyncMap(metrics model.Vector, syncMap *sync.Map) {
+	// create a map of device class to highest osd usage
+	deviceClassUsage := make(map[string]float64)
+	for _, osd := range metrics {
+		deviceClass := osd.Metric["device_class"]
+		if _, ok := deviceClassUsage[string(deviceClass)]; ok {
+			if float64(osd.Value) > deviceClassUsage[string(deviceClass)] {
+				deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+			}
+		} else {
+			deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+		}
+	}
+
+	// update the sync map
+	for deviceClass, usage := range deviceClassUsage {
+		syncMap.Store(deviceClass, usage)
+	}
+}
+
+func getStorageAutoScalerObjects(ctx context.Context, client client.Client) (*ocsv1.StorageAutoScalerList, error) {
+	// get all the storage autoscaler cr objects
+	storageAutoScalerList := &ocsv1.StorageAutoScalerList{}
+	err := client.List(ctx, storageAutoScalerList)
+	if err != nil {
+		return nil, err
+	}
+
+	return storageAutoScalerList, nil
+}
+
+func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList, syncMap *sync.Map, log logr.Logger) []struct {
+	namespace string
+	name      string
+} {
+	var objectList []struct {
+		namespace string
+		name      string
+	}
+
+	for _, storageAutoScaler := range storageAutoScalerList.Items {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		deviceClassThreshold := storageAutoScaler.Spec.StorageScalingThresholdPercent
+		// get the osd usage from the sync map
+		usage, ok := syncMap.Load(deviceClass)
+		if !ok {
+			log.Error(fmt.Errorf("osd usage not found for device class"), "osd usage not found for device class", "device class", deviceClass)
+			continue
+		}
+
+		if usage == nil {
```

Comment:
if the key exists does the `usage` will ever be nil? if that's the case then the issue is at caller end isn't it, since we are the ones setting it. I believe the whole loop can be made dense w/o any continue. possible to give a try?